README
Usepackage Environment Manager
Copyright (C) 1995-2015  Jonathan Hogg  <me@jonathanhogg.com>
   
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful, 
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software 
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


				Usepackage

usepackage is an environment management program. It is based on the principle
of "packages" that are "used". When a package is used, the information
necessary to invoke it is drawn into the environment of the shell. A summary
of how to setup and use the system is given below; see the manual page for 
more information.

Using usepackage:
-----------------

Add one of the following lines to your shell rc file (modify depending on
where you installed usepackage):

 * for csh/tcsh:	source /usr/local/share/usepackage/use.csh
 * for sh/zsh/bash:	source /usr/local/share/usepackage/use.bsh
 * for ksh:		. /usr/local/share/usepackage/use.ksh

Then invoke `use' to add packages to your environment like:

   use <package1> [<package2>...]

For example, in the shell rc file:

   use gnu standard X11

Note that usepackage is case-insensitive, so `x11' is also acceptable in the 
above example.

The `use' command can be called at any time to use another package, for
example:

   % latex foo
   latex: Command not found.
   % use tex
   % latex foo
   This is TeX, Version 3.1415 (C version 6.1)
   ...

If the requested package is not available for your platform then a warning
message will be given of the form:

   % use java
   usepackage: no match for package `java' on this host.

These messages can be suppressed with the `-s' flag (for instance if you log 
onto platforms for which certain packages that you use are not available):

   % use -s java

You can obtain a usage summary by typing `use' with no arguments and a list
of available packages with:

   % use -l

   usepackage 1.9, Copyright (C) 1995-2015  Jonathan Hogg

   Available packages are:

      none - empty paths
      dot - add current directory to PATH (possible security hole)
      [...]

A Recommendation:
-----------------

As usepackage makes it very easy to add packages into your environment at the 
command line, it is worth not adding all the packages you might ever need in 
your shell rc file. Instead add only the packages you commonly use and add 
rarer ones before you run them. For example, if you rarely use latex it makes 
no sense to keep it resident in your path, following the example above it is 
simple to bring it in before you use it. This will keep your path short and 
speed-up shell startup times. In particular, some brain-dead shells have 
major problems with a very long path. Another advantage of doing this is
that the `man' command works much better with a short MANPATH variable and
even better if the pages you request are near the head of the MANPATH (which
is the case when a package is added at the command line).

Customising packages:
---------------------

The file `usepackage.conf' contains the information about each package. This 
file contains comments explaining its format. Per-user packages can be added by 
creating the file `.packages' in your home directory with the same format as 
the main packages file (assuming the `include' line in `usepackage.conf' is
kept). The syntax is also documented in the `usepackage' manual page.


				REPORTING BUGS

If you find a bug in usepackage, please let me know about it. You can email me
at <me@jonathanhogg.com>.



NEWS
Usepackage Environment Manager
Copyright (C) 1995-2020  Jonathan Hogg  <me@jonathanhogg.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Changes between usepackage version 1.13 and 1.14

* Just updating copyrights on everything again for the 25TH ANNIVERSARY!!!
  You may be amazed to hear that I still use this tool on a daily basis.

Changes between usepackage version 1.11 and 1.13

* Just updating the email addresses and copyrights on everything for the
  20TH ANNIVERSARY!!!

Changes between usepackage version 1.11 and 1.12

* More code updating, this time to fix an inclusion bug related to very
  old-school usage of flex.

Changes between usepackage version 1.10 and 1.11

* The code was updated a fair bit to match more modern autoconf/automake
  and C99 standards (it's really starting to show its age).

Changes between usepackage version 1.8 and 1.10

* Now supports shell-independent aliases using 'alias foo = bar' syntax.
* Now supports adding path segments at the end using the new "=+"
  syntax (courtesy of Chandler Wilkerson).
* Skipping 1.9 version number as I never did a real release of it and
  I added the alias stuff afterwards anyway.

Changes between usepackage version 1.7 and 1.8

* Now supports basic checking for existence of directories before adding
  them to a path (or setting a variable) using the new "?=" and "?+="
  operators.
* The order that variables are sourced into the environment has been
  changed to better support using other variables as parts of updates
  to the environment (e.g., "FOO=/opt/foo, PATH+=$FOO/bin;").

Changes between usepackage version 1.6.3 and 1.7

* Now supports arbitrary inclusions into the shell evaluation text,
  allowing commands to be executed, and shell aliases and functions
  to be defined.
* Match patterns (other than the initial package name) are now optional,
  such that a missing pattern defaults to a "*" wildcard.
* An additional optional pattern has been added to match the user's
  shell. This is determined from $SHELL or the shell defined in the
  password database for the running user.

Changes between usepackage version 1.6.2 and 1.6.3

* Fixed ~ expansion to use $HOME if available.

Changes between usepackage version 1.6.1 and 1.6.2

* Split manual page into two (for frontend and backend).
* Cosmetic changes.

Changes between usepackage version 1.6 and 1.6.1

* Now uses autoconf/automake for the build system.
* Small changes to the output messages.
* Fixed up the contact information.
* A bit more portable hopefully.

Initial public release 1.6

* Usepackage has been in use for some time, and had even been available
  for download, but no attempt was made previously to publicise it. This
  release represents the first posted to freshmeat.net.


utils.c

/*****************************************************************************
 * 
 * Usepackage Environment Manager
 * Copyright (C) 1995-2020  Jonathan Hogg  <me@jonathanhogg.com>
 *  
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, 
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software 
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * Name   : utils.c
 * Author : Jonathan Hogg <me@jonathanhogg.com>
 * 
 ****************************************************************************/


/* utils.c */

/* Jonathan AH Hogg */


/*** uses: ***/

#include <pwd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/types.h>
#include "utils.h"


/*** functions: ***/

char* get_user_shell(void)
{
   struct passwd *pwent;
   char *shell = getenv("SHELL");

   if (!shell)
   {
      pwent = getpwuid(getuid());
      shell = pwent->pw_shell;
   }
   shell = strrchr(shell, '/');
   if (!shell)
      return("sh");
   shell++;

   return strdup(shell);
}


char* expand(char* filepath)
{
   static char newpath[256];
   static char username[32];
   struct passwd* pwent;
   char* home;
   int i,j;

   i = 0;
   newpath[0] = '\0';

   if (filepath[i] == '~')
   {
      i++;

      if (!filepath[i] || filepath[i] == '/')
      {
      	 home = getenv("HOME");
      	 if (home)
      	 {
     	    strcat(newpath, home);
      	 }
      	 else
      	 {
	    pwent = getpwuid(getuid());

	    if (!pwent)
	    {
	       fprintf(stderr, "usepackage: cannot obtain home directory.\n");
	       return(filepath);
	    }
	    
	    strcat(newpath, pwent->pw_dir);
	 }
      }
      else
      {
         j = i;
         while (filepath[i] && (filepath[i] != '/')) i++;
         strncpy(username, filepath + j, i - j);
         username[i-j] = '\0';
	 pwent = getpwnam(username);

	 if (!pwent)
	 {
	    fprintf(stderr,
                    "usepackage: cannot obtain home directory of user `%s'.\n",
                    username);
	    return(filepath);
	 }

         strcat(newpath, pwent->pw_dir);
      }
   }

   strcat(newpath, filepath + i);

   return(newpath);
}

